src/tests/system/tests/test_memcache.py

```diff
@@ -1,1660 +1,461 @@
-
-    client.sssd.domain["ldap_id_mapping"] = "false"
-    client.sssd.start()
-
-    check()
     client.sssd.stop()
-    check()
 
-
-@pytest.mark.importance("critical")
-@pytest.mark.cache
-@pytest.mark.topology(KnownTopologyGroup.AnyProvider)
-def test_memcache__lookup_uids_and_check_membership_by_gid(client: Client, provider: GenericProvider):
-    """
-    :title: Lookup user by id and test membership by gid use memory cache when SSSD is stopped
-    :setup:
-        1. Add 'user1', 'user2' and 'user3' to SSSD
-        2. Set users uids and gids
-        3. Add 'group1' and 'group2' to SSSD
-        4. Set groups gids
-        5. Add users to groups
-        6. Start SSSD
-    :steps:
-        1. Find users by id(uid)
-        2. Check that users are members of correct groups
-        3. Stop SSSD
-        4. Find users by id(uid)
-        5. Check that users are members of correct groups
-    :expectedresults:
-        1. Users are found
-        2. Users are members of correct groups
-        3. SSSD is stopped
-        4. Users are found
-        5. Users are members of correct groups
-    :customerscenario: False
-    """
-
-    def check():
-        result = client.tools.id(2001)
-        assert result is not None, "User with id 2001 was not found using id"
-        assert result.memberof([101, 1001, 1002]), "User with id 2001 is member of incorrect groups"
-
-        result = client.tools.id(2002)
-        assert result is not None, "User with id 2002 was not found using id"
-        assert result.memberof([102, 1002]), "User with id 2002 is member of incorrect groups"
-
-        result = client.tools.id(2003)
-        assert result is not None, "User with id 2003 was not found using id"
-        assert result.memberof([103, 1002]), "User with id 2003 is member of incorrect groups"
-
-    u1 = provider.user("user1").add(uid=2001, gid=101)
-    u2 = provider.user("user2").add(uid=2002, gid=102)
-    u3 = provider.user("user3").add(uid=2003, gid=103)
-
-    provider.group("group1").add(gid=1001).add_member(u1)
-    provider.group("group2").add(gid=1002).add_members([u1, u2, u3])
-
-    client.sssd.domain["ldap_id_mapping"] = "false"
-    client.sssd.start()
-
-    check()
-    client.sssd.stop()
-    check()
+    if memcache:
+        if lookup_type in ("users"):
+            assert (
+                client.tools.getent.passwd("user1") is not None
+            ), "`user1` should still be found after stopping SSSD."
+        if lookup_type in ("groups"):
+            assert (
+                client.tools.getent.group("group1") is not None
+            ), "`group1` should still be found after stopping SSSD."
+    elif not memcache:
+        if lookup_type in ("users"):
+            assert client.tools.getent.passwd("user1") is None, "`user1` should NOT be found after stopping SSSD."
+        if lookup_type in ("groups"):
+            assert client.tools.getent.group("group1") is None, "`group1` should NOT be found after stopping SSSD."
```

Comment:
This is still unsightly. How about rewriting it like this (maybe parametrize only "lookup_type", ["users", "groups"]?
1. sssd is running and user and group is found.
2. sssd is stopped with memcache and user and group is found.
3. sssd is started again without memcache and user and group is found.
4. sssd is stopped without memcache and user and group is not found.

If You make it in parametrized to 4 tests it will be all of the setup and teardown will be run 4 times for each of the provider making the overhead costly and the test less readable. It might be better to have more steps it the test instead at the cost that firs failure will stop the test.

Reply:
There are indeed pros and cons following the approach you suggested. Figuring it out took longer than getting it done.

My opinion is that, the setup and teardown are not so worth it when compared to start/stopping SSSD multiple times inside a single test; I would prefer to have 4 different tests and it telling me where is the failure rather than trying to guess if there are multiple failures when the test fails.